File: 3235-minimum-cost-to-convert-string-i/minimum-cost-to-convert-string-i.cpp
class Solution {
public:
    using ll=long long;
    long long minimumCost(string source, string target, vector<char>& original, vector<char>& changed, vector<int>& cost) {
        ios_base::sync_with_stdio(0);
        const int INF=(int)1e7;
        vector<vector<int>> costs(26,vector<int> (26,INF));
        for(int i=0;i<cost.size();i++){
            int t1=original[i]-'a';
            int t2=changed[i]-'a';
            costs[t1][t2]=min(costs[t1][t2],cost[i]);
        }

        for(int k=0;k<26;k++){
            for(int i=0;i<26;i++){
                for(int j=0;j<26;j++){
                    if(i==j) continue;
                    if(costs[i][j]>costs[i][k]+costs[k][j])
                        costs[i][j]=costs[i][k]+costs[k][j];
                }
            }
        }
        ll ans=0;
        for(int i=0;i<source.size();i++){
            if(source[i]==target[i]) continue;
            int c1=source[i]-'a';
            int c2=target[i]-'a';
            if(costs[c1][c2]==INF) return -1;
            ans+=costs[c1][c2];
        }   
        return ans;
    }
};

File: 3235-minimum-cost-to-convert-string-i/minimum-cost-to-convert-string-i.py
INF=int(1e13+7)

class Solution:
    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        costDict=[[INF for _ in range(26)] for _ in range(26)]
        for i in range(len(original)):
            c=original[i]
            d=changed[i]
            costT=cost[i]
            cI=ord(c)-ord('a')
            dI=ord(d)-ord('a')
            costDict[cI][dI]=min(costT,costDict[cI][dI])
        #print(costDict)
        
        for i in range(26):
            costDict[i][i]=0

        for k in range(26):
            for i in range(26):
                for j in range(26):
                    if costDict[i][k]+costDict[k][j]<costDict[i][j]:
                        costDict[i][j]=costDict[i][k]+costDict[k][j]

        total=0
        for x in range(len(source)):
            c=source[x]
            cT=target[x]
            i=ord(c)-ord('a')
            iT=ord(cT)-ord('a')
            # print(f"{c} to {cT} in {dp[i][iT]}")
            if(costDict[i][iT]==INF):
                return -1
            else:
                total+=costDict[i][iT]
        # for row in dp:
        #     for i in row:
        #         print(i if i!=INF else 'X',end="   ")
        #     print()
        return total